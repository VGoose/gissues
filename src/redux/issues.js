import { createAction, createReducer } from 'redux-starter-kit'
import axios from '../utils/axios'

//export for testing
export const getIssuesRequest = createAction('getIssuesRequest')
export const getIssuesSuccess = createAction('getIssuesSuccess')
export const getIssuesError = createAction('getIssuesError')
export const setRepo = createAction('setRepo')

export const getIssues = (owner, repo) => (dispatch, getState) => {
  dispatch(getIssuesRequest())
  dispatch(setRepo(repo))
  //use cache or make get req
  const cachedIssues = getIssuesFromStorage(repo)
  cachedIssues
    ? dispatch(getIssuesSuccess(cachedIssues))
    : axios.get(`repos/${owner}/${repo}/issues?access_token=${getState().key.key}`)
      .then(res => dispatch(getIssuesSuccess(res.data)))
      .catch(e => dispatch(getIssuesError(e)))
}
export const updateIssues = (issues) => (dispatch, getState) => {
  //update state and cache
  dispatch(getIssuesSuccess(issues))
  storeIssues(getState().issues.repo, issues)
}

//side effects
const storeIssues = (repo, issues) => {
  localStorage.setItem(repo, JSON.stringify(issues))
}
const getIssuesFromStorage = (repo) => {
  const value = localStorage.getItem(repo)
  return value && JSON.parse(value)
}

const initialState = {
  isLoading: false, 
  issues: [], 
  repo: null, 
  error: null
}
const issuesReducer = createReducer(initialState, {
  [setRepo]: (state, action) => { state.repo = action.payload },
  [getIssuesRequest]: (state) => { state.isLoading = true },
  [getIssuesSuccess]: (state, action) => {
    state.isLoading = false
    state.issues = action.payload
  },
  [getIssuesError]: (state, action) => {
    state.isLoading = false
    state.error = action.payload
  }
})

export default issuesReducer