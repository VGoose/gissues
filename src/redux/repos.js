import { createAction, createReducer } from 'redux-starter-kit'

//export for testing
export const getReposRequest = createAction('getReposRequest')
export const getReposSuccess = createAction('getReposSuccess')
export const getReposError = createAction('getReposError')

const initialState = {
  isLoading: false,
  repos: [],
  error: null
}
const reposReducer = createReducer(initialState, {
  [getReposRequest]: (state) => { state.isLoading = true },
  [getReposSuccess]: (state, action) => {
    state.isLoading = false
    state.repos = action.payload
  },
  [getReposError]: (state, action) => {
    state.isLoading = false
    state.error = action.payload
  }
})

export default reposReducer