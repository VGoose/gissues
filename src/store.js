import { configureStore } from 'redux-starter-kit'

import rootReducer from './redux'

const store = configureStore({
  reducer: rootReducer
})

export default store