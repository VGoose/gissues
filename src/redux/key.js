import { createAction, createReducer } from 'redux-starter-kit'
import axios from '../utils/axios'
import { getReposSuccess } from './repos'

//export for testing
export const validateKeyRequest = createAction('validateKeyRequest')
export const validateKeySuccess = createAction('validateKeySuccess')
export const validateKeyError = createAction('validateKeyError')

export const validateKey = (key) => dispatch => {
  dispatch(validateKeyRequest())
  axios.get(`user/repos?access_token=${key}`)
    .then(res => {
      dispatch(getReposSuccess(res.data))
      dispatch(validateKeySuccess(key))
      dispatch(validateKeyError(null))
      storeKey(key)
    })
    .catch(e => dispatch(validateKeyError(e)))
}

export const invalidateKey = () => dispatch => {
  dispatch(validateKeySuccess(null))
  localStorage.removeItem('key')
}

const storeKey = (key) => {
  localStorage.setItem('key', key)
}
export const getKeyFromStorage = () => dispatch => {
  const value = localStorage.getItem('key')
  if (value) {
    dispatch(validateKey(value))
  } else {
    dispatch(validateKeyError(null))
  }
}
const initialState = { 
  isLoading: true, //buy time to check auth on refresh - prevent flashing
  key: null, 
  error: null 
}
const keyReducer = createReducer(initialState, {
  [validateKeyRequest]: (state) => { state.isLoading = true },
  [validateKeySuccess]: (state, action) => {
    state.isLoading = false
    state.key = action.payload
  },
  [validateKeyError]: (state, action) => {
    state.isLoading = false
    state.error = action.payload
  }
})

export default keyReducer