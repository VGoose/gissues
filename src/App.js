import React from 'react'
import styled, { createGlobalStyle } from 'styled-components'
import { connect } from 'react-redux'
import KeyForm from './components/key_form.js'
import ReposViewer from './components/repos_viewer'
import Button from './components/button.js'
import { invalidateKey, getKeyFromStorage } from './redux/key.js'

//export for testing
export const App = ({ authenticated, dispatch, keyIsLoading }) => {
  //check if authed on mount
  React.useEffect(() => {
    dispatch(getKeyFromStorage())
  }, [])
  return (
    <Wrapper className="App">
      <GlobalStyle />
      < Header > Gissues</Header >
      {
        authenticated
          ? <React.Fragment >
            <Description>Click on an repo to view issues!</Description>
            <Button onClick={() => dispatch(invalidateKey())}>Use new Key</Button>
            <ReposViewer />
          </React.Fragment >
          
          : keyIsLoading
            ? <div></div>
            : <KeyForm />}
    </Wrapper >
  )
}

const GlobalStyle = createGlobalStyle`
  * {
        box-sizing: border-box;
    }
  `

const Wrapper = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    max-width: 900px;
    margin: 0 auto;
    padding: 0 2rem;
  `
const Header = styled.h1`
    font-size: 4rem;
    color: #FF5313;
  `
const Description = styled.p`
  
  `
const mapStateToProps = state => ({
  authenticated: state.key.key,
  keyIsLoading: state.key.isLoading,
})

export default connect(mapStateToProps, null)(App)
