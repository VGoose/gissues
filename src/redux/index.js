import { combineReducers } from 'redux'

import issues from './issues'
import key from './key'
import repos from './repos'

const rootReducer = combineReducers({
  key,
  repos,
  issues
})

export default rootReducer


