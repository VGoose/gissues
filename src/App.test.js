import React from 'react'
import App from './App'
import renderer from 'react-test-renderer'
import store from './store'
import { Provider } from 'react-redux'

import { validateKeyError, validateKeySuccess } from './redux/key'


//renderer.create doesn't like nested components? 
const defaultStore = {...store}
it('first visit render tree', () => {
  const MyApp = () => <Provider store={defaultStore}><App /></Provider>
  const tree = renderer.create(<MyApp />).toJSON()
  expect(tree).toMatchSnapshot()
})

// const store2 = {...store}
// store2.dispatch(validateKeyError('err'))
// const MyApp2 = () => <Provider store={store2}><App /></Provider>
// it('is not authed render tree', () => {
//   const tree = renderer.create(<MyApp2 />).toJSON()
//   expect(tree).toMatchSnapshot()
// })
// store.dispatch(validateKeySuccess('foo'))
// const MyApp3 = () => <Provider store={store}><App /></Provider>
// it('is authed render tree', () => {
//   const tree = renderer.create(<MyApp />).toJSON()
//   expect(tree).toMatchSnapshot()
// })