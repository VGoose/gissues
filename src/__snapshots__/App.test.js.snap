// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first visit render tree 1`] = `
<div
  className="App sc-gZMcBi ivEedh"
>
  <h1
    className="sc-gqjmRU kngvag"
  >
     Gissues
  </h1>
  <div />
</div>
`;
